event-web-framework/draft.markdown

```diff
@@ -0,0 +1,941 @@
+
+```lisp
+(defmethod buffer! ((buffer buffer))
+  (handler-case
+      (let ((stream (bi-stream buffer)))
+    	(incf (tries buffer))
+    	(loop for char = (read-char-no-hang stream) until (null char)
+    	   do (push char (contents buffer))
+    	   do (incf (total-buffered buffer))
+    	   when (request buffer) do (decf (expecting buffer))
+    	   when (line-terminated? (contents buffer))
+    	   do (multiple-value-bind (parsed expecting) (parse buffer)
+    		(setf (request buffer) parsed
+    		      (expecting buffer) expecting)
+    		(return char))
+    	   when (> (total-buffered buffer) +max-request-size+) return char
+    	   finally (return char)))
+    (error () :eof)))
+```
+
+When `buffer!` is called on a `buffer`, it:
+- increments the `tries` count, so that we can evict "needy" buffers in `process-ready`;
+- loops to read characters from the input stream, and
+- returns the last character it read if it has read all of the available input.
+- It also tracks any `\r\n\r\n` sequences so that we can later detect complete requests.
+- Finally, any error results it returns an `:eof` to signal that `process-ready` should discard this particular connection.
+
+The `buffer` type is a CLOS [_class_](http://www.gigamonkeys.com/book/object-reorientation-classes.html). Classes in CLOS let us define a type with fields called `slots`. We don't see the behaviours associated with `buffer` on the class definition, because (as we've already learned), we do that using generic functions like `buffer!`.
+
+`defclass` does allow us to specify getters/setters (`reader`s/`accessor`s), and slot initializers; `:initform` specifies a default value, while `:initarg` identifies a hook that the caller of `make-instance` can use to provide a default value.
+
+```lisp
+(defclass buffer ()
+  ((tries :accessor tries :initform 0)
+   (contents :accessor contents :initform nil)
+   (bi-stream :reader bi-stream :initarg :bi-stream)
+   (total-buffered :accessor total-buffered :initform 0)
+   (started :reader started :initform (get-universal-time))
+   (request :accessor request :initform nil)
+   (expecting :accessor expecting :initform 0)))
+```
+
+Our `buffer` class has six slots:
+
+- `tries`, which keeps count of how many times we've tried reading into this buffer
+- `contents`, which contains what we've read so far
+- `bi-stream`, which a hack around some of those Common Lisp-specific, non-blocking-IO annoyances I mentioned earlier
+- `total-buffered`, which is a count of chars we've read so far
+- `started`, which is a timestamp that tells us when we created this buffer
+- `request`, which will eventually contain the request we construct from buffered data
+- `expecting`, which will signal how many more chars we're expecting (if any) after we buffer the request headers
```

Comment:
I ended up not rendering these as tables, because bulleted list was easier to read in both web and print/PDF formats.